partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelOpWindow.kt

```diff
@@ -0,0 +1,165 @@
+    private val input: ExprRelation,
+    private val functions: List<WindowFunction>,
+    private val partitionBy: List<ExprValue>,
+    private val sortBy: List<Collation>
+) : RelOpPeeking() {
+
+    private companion object {
+        private val comparator = DatumArrayComparator
+    }
+
+    private lateinit var _env: Environment
+
+    /**
+     * This is used to track the current partition that we are iterating through.
+     */
+    private var _partition: LocalPartition = LocalPartition()
+
+    /**
+     * This is used to track where we are in the partition that we are currently iterating through.
+     * @see peek
+     */
+    private var _partitionPeekingNumber: Long = 0
+
+    /**
+     * When lazily creating the partition, we need to step out-of-bounds from the current partition to know
+     * whether we are in a new partition. So, we need to put the out-of-bounds row into a place to be used
+     * when creating the next partition.
+     */
+    private var leftoverRow: Row? = null
+
+    override fun openPeeking(env: Environment) {
+        input.open(env)
+        this._env = env
+        _partitionPeekingNumber = -1L
+        _partition = LocalPartition()
+        functions.map { it.reset(_partition) }
+    }
+
+    override fun peek(): Row? {
+        // Check if there is an existing partition. If so, evaluate and return.
+        _partitionPeekingNumber++
+        if (_partition.size() > _partitionPeekingNumber) {
+            return produceResult()
+        }
+        _partitionPeekingNumber = 0L
+
+        // Create new partition's first row
+        var partitionCreationIndex = 0L
+        val newLocalPartition = LocalPartition()
+        val firstRow = when {
+            leftoverRow != null -> {
+                val tempRow = leftoverRow!!
+                leftoverRow = null
+                tempRow
+            }
+            else -> when (input.hasNext()) {
+                true -> input.next()
+                false -> return null
+            }
+        }
+        var previousInfoIndex = newLocalPartition.add(OrderingInfo(partitionCreationIndex))
+        newLocalPartition.add(firstRow, previousInfoIndex)
+        val newEnv = _env.push(firstRow)
+        val firstRowPartitionKeys = Array(partitionBy.size) { partitionBy[it].eval(newEnv) }
+        var previousRowSortKeys = Array(sortBy.size) { sortBy[it].expr.eval(newEnv) }
+
+        // Add partition's remaining rows
+        while (input.hasNext()) {
+            partitionCreationIndex++
+            val nextRow = input.next()
+            val nextEnv = _env.push(nextRow)
+
+            // Stop (and save spillover row) if at next partition
+            val nextPartitionKeys = Array(partitionBy.size) { partitionBy[it].eval(nextEnv) }
+            val isNewPartition = comparator.compare(firstRowPartitionKeys, nextPartitionKeys) != 0
+            if (isNewPartition) {
+                leftoverRow = nextRow
+                break
+            }
```

Comment:
Notes: we reached the end of partition. Since we have already evaluated the first row of next partition, cache it.